Implement _WKSerializedNode of DocumentFragment, template elements, and shadow roots

achristensen07 opened the pull request:
#### 68b4ea83bc8107bb5e0907815d9982c42a09cde1
<pre>
Implement _WKSerializedNode of DocumentFragment, template elements, and shadow roots
<a href="https://bugs.webkit.org/show_bug.cgi?id=298367">https://bugs.webkit.org/show_bug.cgi?id=298367</a>
<a href="https://rdar.apple.com/159811421">rdar://159811421</a>

Reviewed by Ryosuke Niwa.

This makes _WKSerializedNode able to handle anything that Node.cloneNode can
except the interactions with CustomElementRegistry, which is a little unclear because
Node.cloneNode only works within the scope of one document but _WKSerializedNode doesn&apos;t.

I use a little bit of explicit template instantiation so I can return nested types without
including SerializedNode.h from other headers to keep the build time low.

* Source/WebCore/dom/DocumentFragment.cpp:
(WebCore::DocumentFragment::serializeNode const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::serializeShadowRoot const):
(WebCore::Element::serializeAttributes const):
(WebCore::Element::serializeNode const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/SerializedNode.cpp:
(WebCore::addShadowRootIfNecessary):
(WebCore::SerializedNode::deserialize):
* Source/WebCore/dom/SerializedNode.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::serializeNode const):
* Source/WebCore/dom/SlotAssignmentMode.h:
* Source/WebCore/dom/TemplateContentDocumentFragment.h:
* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::adoptDeserializedContent):
(WebCore::HTMLTemplateElement::serializeNode const):
* Source/WebCore/html/HTMLTemplateElement.h:
* Source/WebKit/Shared/SerializedNode.serialization.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SerializedNode.mm:
(TestWebKitAPI::TEST(SerializedNode, Basic)):

Canonical link: <a href="https://commits.webkit.org/299731@main">https://commits.webkit.org/299731@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/936ff65590963cb130ddecd53eafa711d36b7060

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119975 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39668 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30319 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126305 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72048 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/de406489-8e62-4094-b51c-1858ca61f2c2) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121851 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40364 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48245 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91110 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60421 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5d879bce-b809-4287-88db-8289ab4d1b97) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122927 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32241 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107579 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71666 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/121565da-0387-460b-98de-30934d4f5912) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31272 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25684 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69940 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101716 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25872 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129225 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46895 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35557 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99731 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47261 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103767 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99576 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45026 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23043 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43516 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19081 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46757 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52463 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46223 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49572 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47909 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/ShadowRoot.cpp:
```diff
@@ -304,10 +304,26 @@ Ref<Node> ShadowRoot::cloneNodeInternal(Document& document, CloningOperation typ
+        break;
+    }
+
+    RELEASE_ASSERT_NOT_REACHED(); // ShadowRoot is never serialized directly on its own.
```
Can't someone call `serializeNode` on `ShadowRoot` directly though?

rniwa replied:
There is no check in `WebKitNamespace::serializeNode` or `WebKitSerializedNode::WebKitSerializedNode` that prevent `serializeNode` to be called on `ShadowRoot` directly. We should probably add some kind of check in either function if we're going to release-assert like here.

achristensen07 replied:
I'd like to keep WebKitNamespace::serializeNode the same shape as Node::cloneNodeForBindings, so I'll add the same check.

rniwa replied:
Sounds good.

rniwa commented on Source/WebCore/html/HTMLTemplateElement.cpp:
```diff
@@ -141,11 +146,17 @@ SerializedNode HTMLTemplateElement::serializeNode(CloningOperation type) const
-
-    return { SerializedNode::HTMLTemplateElement { { { WTFMove(children) }, { tagQName() }, WTFMove(attributes) } } };
+    RefPtr content = m_content;
+    auto contentChildren = content ? std::optional(SerializedNode::DocumentFragment { content->serializeChildNodes() }) : std::nullopt;
```
Here we're not respecting the distinction between `SelfOnly` and `SelfWithTemplateContent`. In that case, maybe better to define a new enum for `serializeNode` with just two values.

achristensen07 replied:
This was an oversight.  I'm going to respect the difference just like cloneNode does.

rniwa commented on Source/WebCore/dom/ShadowRoot.cpp:
```diff
@@ -304,10 +304,26 @@ Ref<Node> ShadowRoot::cloneNodeInternal(Document& document, CloningOperation typ
+    ASSERT(m_isClonable);
+    switch (type) {
+    case CloningOperation::SelfWithTemplateContent:
+        return { SerializedNode::ShadowRoot { { serializeChildNodes() },
```
We also need to serialize adopted stylesheets (`adoptedStyleSheets()` in `TreeScope`) for `ShadowRoot` and `Document` but it's okay to do that in a follow up.

rniwa commented on Source/WebCore/dom/SerializedNode.h:
```diff
@@ -72,27 +73,31 @@ struct SerializedNode {
         String systemId;
     };
+    struct ShadowRoot : public DocumentFragment {
+        bool openMode;
```
Instead of using a bunch of bools like this, we should use enum classes defined in ShadowRoot.h (maybe move them out of `ShadowRoot` so that it can be forward declared).
Otherwise there is no type enforcement so bools can swap positions without us noticing.

achristensen07 replied:
ShadowRootMode::UserAgent isn't cloned or serialized, so that needs to be something else.  But I can use existing ones or move them and do this for the rest.

rniwa reviewed (approved):
Looks okay modulo issues I pointed out above.

achristensen07 committed:
Implement _WKSerializedNode of DocumentFragment, template elements, and shadow roots
https://bugs.webkit.org/show_bug.cgi?id=298367
rdar://159811421

Reviewed by Ryosuke Niwa.

This makes _WKSerializedNode able to handle anything that Node.cloneNode can
except the interactions with CustomElementRegistry, which is a little unclear because
Node.cloneNode only works within the scope of one document but _WKSerializedNode doesn't.

I use a little bit of explicit template instantiation so I can return nested types without
including SerializedNode.h from other headers to keep the build time low.

* Source/WebCore/dom/DocumentFragment.cpp:
(WebCore::DocumentFragment::serializeNode const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::serializeShadowRoot const):
(WebCore::Element::serializeAttributes const):
(WebCore::Element::serializeNode const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/SerializedNode.cpp:
(WebCore::addShadowRootIfNecessary):
(WebCore::SerializedNode::deserialize):
* Source/WebCore/dom/SerializedNode.h:
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::serializeNode const):
* Source/WebCore/dom/SlotAssignmentMode.h:
* Source/WebCore/dom/TemplateContentDocumentFragment.h:
* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::adoptDeserializedContent):
(WebCore::HTMLTemplateElement::serializeNode const):
* Source/WebCore/html/HTMLTemplateElement.h:
* Source/WebKit/Shared/SerializedNode.serialization.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SerializedNode.mm:
(TestWebKitAPI::TEST(SerializedNode, Basic)):

Canonical link: https://commits.webkit.org/299731@main
